docs/proposals/petset.md

```diff
@@ -0,0 +1,414 @@
+  than running a clustered system off Kube. Authors should be given tools to guard against common cluster
+  failure modes (split brain, phantom member) to prevent introducing more failure modes. Sophisticated
+  distributed systems designers can implement more sophisticated solutions than PetSet if necessary -
+  new users should not become vulnerable to additional failure modes through an overly flexible design.
+* **Limited scaling** - While flexible scaling is important for some clusters, other examples of clusters
+  do not change scale without significant external intervention. Human intervention may be required after
+  scaling. Changing scale during cluster operation can lead to split brain in quorum systems. It should be
+  possible to scale easily, but the details of making that safe belong to the pods and image authors.
+* **No generic cluster lifecycle** - Rather than design a general purpose lifecycle for clustered software,
+  focus on ensuring the information necessary for the software to function is available. For example,
+  rather than providing a "post-creation" hook invoked when the cluster is complete, provide the necessary
+  information to the "first" (or last) pod to determine the identity of the remaining cluster members and
+  allow it to manage its own initialization.
+* **External access direct to cluster members is out of scope** - exposing pods to consumers that cannot
+  access the pod network is out of scope, because we do not currently support headless services being
+  exposed via NodePort. A workaround is to allow external clients to access the pod network, or to
+  create one NodePort service per member. A future design should cover headless external service access.
+
+
+## Proposed Design
+
+Add a new resource to Kubernetes to represent a set of pods that are individually distinct but each
+individual can safely be replaced-- the name **PetSet** (working name) is chosen to convey that the
+individual members of the set are themselves "pets" and thus each one is preserved. A relevant analogy
+is that a PetSet is composed of pets, but the pets are like goldfish. If you have a blue, red, and
+yellow goldfish, and the red goldfish dies, you replace it with another red goldfish and no one would
+notice. If you suddenly have three red goldfish, someone will notice.
+
+The PetSet is responsible for creating and maintaining a set of **identities** and ensuring that there is
+one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
+or unique supporting resource per identity at any one time. A new pod can be created for an identity only
+if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
+
+A PetSet has 0..N **members**, each with a unique **identity** which is a name that is unique within the
+set.
+
+```
+type PetSet struct {
+    ObjectMeta
+
+    Spec PetSetSpec
+    ...
+}
```

Comment:
This replaces your `MemberIdentities` with just a template. Can you counter? 

```yaml
kind: PetSet
metadata:
  name: mysql-master
spec:
  template: mysql-master
  replicas: 1
  parameters:
  - name: MASTER_LABEL
    value: master
  - name: MASTER_PVC
    value: mysql-master-pvc
---
kind: PetSet
metadata:
    name: mysql-slave
spec:
  template: mysql-slave
  replicas: 5
  parameters:
  - name: SLAVE_LABEL
    value: slave
---
kind: Template
metadata:
  name: mysql-master
objects:
  # RC, or some variation thereof
  - kind: ReplicationController
    metadata:
      labels:
        app: $(MASTER_LABEL)
      name: $(MASTER_NAME)
    spec:
      replicas: 1
      template:
        metadata:
          labels:
            app: $(MASTER_LABEL)
        spec:
          containers:
          - image: mysql-master
            name: master
            volumeMounts:
            - name: datadir
              mountPath: "/var/lib/mysql"
        volumes:
        - name: datadir
          persistentVolumeClaim:
            claimName: $(MASTER_PVC)
  - kind: Service
    metadata:
      name: mysql-master
      labels:
        name: mysql-master
    spec:
      ports:
      - port: 3306
      selector:
        name: mysql-master
parameters : {}
---
kind: Template
metadata:
  name: mysql-slave
objects:
  - kind: ReplicationController
    metadata:
      labels:
        app: $(SLAVE_LABEL)
      # ID is populated by petset controller, per its "replicas"
      name: mysql-slave-$(ID)
    spec:
      replicas: 1
      template:
        metadata:
          labels:
            app: $(SLAVE_LABEL)
        spec:
          containers:
          - image: mysql-slave
            name: slave$(ID)
            volumeMounts:
            - name: datadir
              mountPath: "/var/lib/mysql"
        volumes:
        - name: datadir
          persistentVolumeClaim:
            claimName: mysql-slave-$(ID)
parameters : {}
```

Then all the petset controller needs to do is 
1. populate fields in the pointed at template
2. responsibly assign ID according to its `replicas:`

It doesn't allow nesting like your proposal, but maybe we can converge (feel like that's a little more advanced)?

I also think we'll need a way to express different hooks for master vs slave, so having a petset for each helps.